Internal `repos-config` new syntax

rjbou opened the pull request:
It is quite difficult and misleading to repository related information in the `repos-config` file (like in #4933 and #6283).
The current format contain a list with a some optional elements, which complicated the parsing
> `repositories: [ <string> { <URL> } { {<string>}+ } { <int> } ... ]`: lists the configured repository idents, their URLs and trust anchors. The format is similar to [repositories:](https://opam.ocaml.org/doc/Manual.html#opamrcfield-repositories) from opamrc, except that the <URL> itself is optional.

The PR implements a new syntax for this file: each repository is a section with it's own defined labeled fields:
```
repo <string> {
  url: <URL>
  fingerprints:  <string>
  quorum: <int>
}
```
Adding then a new field is quite easy.

This change implies an hard upgrade (no best effort loading if read only) to update the opam root.

fix #6327

TODO:
* [ ] update doc
* [ ] update changes
* [x] update opam root test
* [x] move to light upgrade
* [x] add a comment on hard upgrade that it should not be changed, ideally (done in #6417)
* [x] queued on #6417 
* [x] queued on #6839

kit-ty-kate commented on src/state/opamFormatUpgrade.ml:
```diff
@@ -1147,6 +1150,32 @@ let v2_2 = OpamVersion.of_string "2.2"
+      in
+      OpamFile.Repos_config.write f repos)
+    (OpamFile.Repos_config_Legacy.BestEffort.read_opt
+       (OpamFile.make
+          (OpamFile.filename f)));
```
```suggestion
       (OpamFile.make (OpamFile.filename f)));
```

kit-ty-kate commented on tests/reftests/repository.test:
```diff
@@ -838,34 +838,34 @@ nourl        file://${BASEDIR}/nourl nourl
 [WARNING] Errors in ${BASEDIR}/OPAM/repo/repos-config, some fields have been ignored:
-            - At ${BASEDIR}/OPAM/repo/repos-config:2:16-2:23::
-              expected url
+            - In ${BASEDIR}/OPAM/repo/repos-config:
```
what is the reason we're loosing the position here?

rjbou replied:
OpamFormat.section does not keep positions

kit-ty-kate reviewed (commented):
I'd be interested in a crash course on `OpamPp` at the next meeting to be able to understand the new `Repos_configSyntax` module better but overall otherwise that looks fine to me.

We might also want to discuss removing the soft/hard upgrade distinction.

kit-ty-kate commented on src/format/opamFile.ml:
```diff
@@ -1386,7 +1386,7 @@ module ConfigSyntax = struct
   let format_version = OpamVersion.of_string "2.1"
   let file_format_version = OpamVersion.of_string "2.0"
-  let root_version = OpamVersion.of_string "2.2"
+  let root_version = OpamVersion.of_string "2.4"
```
```suggestion
  let root_version = OpamVersion.of_string "2.4~alpha1"
```

rjbou commented:
New implem, simpler, and that permit to introduce repository variables & repositories config option update via `opam option`/`opam var`.

rjbou commented on tests/reftests/repository.test:
```diff
@@ -824,37 +824,157 @@ nourl        file://${BASEDIR}/nourl nourl
+### # ---
+### <OPAM/repo/repos-config>
+opam-version: "2.0"
+### opam repo --all
```
It seems that we can have no repo configured...

rjbou replied:
related to #6816

kit-ty-kate commented on src/client/opamClient.ml:
```diff
@@ -1849,9 +1849,16 @@ let init
-          | Some r -> [r.repo_name, (r.repo_url, r.repo_trust)]
-          | None -> OpamFile.InitConfig.repositories init_config
+        let repos =
+          let open OpamFile.Repos_config in
```
This is not necessary and is causing an error for unused open
```suggestion
```

kit-ty-kate commented on src/client/opamClient.ml:
```diff
@@ -1849,9 +1849,16 @@ let init
+          | None ->
+            List.map (fun (n,(u,t)) ->
+                n, OpamFile.Repo_config.create ?trust:t u)
+              ( OpamFile.InitConfig.repositories init_config)
```
```suggestion
              (OpamFile.InitConfig.repositories init_config)
```

kit-ty-kate commented on src/format/opamFile.ml:
```diff
@@ -1975,6 +1981,141 @@ module Repos_configSyntax = struct
+
+  let empty = {
+    url = OpamUrl.empty;
+    trust  = None;
```
```suggestion
    trust = None;
```

kit-ty-kate commented on src/format/opamFile.ml:
```diff
@@ -1370,7 +1370,7 @@ module ConfigSyntax = struct
   let format_version = OpamVersion.of_string "2.1"
   let file_format_version = OpamVersion.of_string "2.0"
-  let root_version = OpamVersion.of_string "2.2"
+  let root_version = OpamVersion.of_string "2.4~alpha1"
```
```suggestion
  let root_version = OpamVersion.of_string "2.5~alpha1"
```

kit-ty-kate commented on src/format/opamFile.ml:
```diff
@@ -1975,6 +1981,141 @@ module Repos_configSyntax = struct
+
+module Repo_configSyntax = struct
+  let internal = "repo-config"
+  let format_version = OpamVersion.of_string "2.1"
```
```suggestion
  let format_version = OpamVersion.of_string "2.5"
```

kit-ty-kate commented on src/state/opamFormatUpgrade.ml:
```diff
@@ -1147,15 +1150,47 @@ let v2_2 = OpamVersion.of_string "2.2"
 
 let from_2_2_beta_to_2_2 ~on_the_fly:_ _ conf = conf, gtc_none
 
+let v2_3 = OpamVersion.of_string "2.3"
+
+let from_2_2_to_2_3 ~on_the_fly:_ _ conf = conf, gtc_none
+
```
i don't think it is

kit-ty-kate commented on src/state/opamFormatUpgrade.ml:
```diff
@@ -1152,9 +1156,41 @@ let v2_2 = OpamVersion.of_string "2.2"
+
+let from_2_2_to_2_3 ~on_the_fly:_ _ conf = conf, gtc_none
+
+let v2_4_alpha1 = OpamVersion.of_string "2.4~alpha1"
```
```suggestion
let v2_5_alpha1 = OpamVersion.of_string "2.5~alpha1"
```

kit-ty-kate commented on src/state/opamFormatUpgrade.ml:
```diff
@@ -1152,9 +1156,41 @@ let v2_2 = OpamVersion.of_string "2.2"
+
+let v2_4_alpha1 = OpamVersion.of_string "2.4~alpha1"
+
+let from_2_3_to_2_4_alpha1_repo ?config:_ root _conf =
```
```suggestion
let from_2_2_to_2_5_alpha1_repo ?config:_ root _conf =
```

kit-ty-kate commented on src/state/opamFormatUpgrade.ml:
```diff
@@ -1152,9 +1156,41 @@ let v2_2 = OpamVersion.of_string "2.2"
+    (OpamFile.Repos_config_Legacy.BestEffort.read_opt
+       (OpamFile.make (OpamFile.filename f)))
+
+let from_2_3_to_2_4_alpha1 ~on_the_fly root conf =
```
```suggestion
let from_2_2_to_2_5_alpha1 ~on_the_fly root conf =
```

kit-ty-kate reviewed (approved):
lgtm modulo history cleanup and the handful of comments

rjbou committed:
Add some documentation in format upgrade about repo & switch upgrades

rjbou committed:
Change repos-config file format

rjbou committed:
opam root version: make seds on root version more robust

rjbou committed:
global state: add global lock to the record in order to retrieve it when needed

rjbou committed:
format upgrade: Use already created global lock instead of recreating one

Recreating the lock reopened the fd on the lock file causing a block when the write lock was taken.

kit-ty-kate commented on src/format/opamFile.ml:
```diff
@@ -2000,6 +2002,145 @@ module Repos_configSyntax = struct
+  let with_repos repos t = { t with repos }
+
+  let fields = [
+    "opam-version", Pp.ppacc with_opam_version opam_version
```
shouldn't opam-version kept at 2.0 but a new field be added similar to opam-root-version in /config?
